Support PHP 8

deleugpn opened the pull request:
Provide PHP 8 Support.

GrahamCampbell commented on composer.json:
```diff
@@ -15,11 +15,11 @@
     },
     "require-dev": {
         "jeremeamia/superclosure": "^2.0",
-        "phpunit/phpunit": "^4.0 || ^5.0 || ^6.0 || ^7.0"
```
Need to keep these in,

GrahamCampbell commented on tests/SignedClosureTest.php:
```diff
@@ -119,11 +115,10 @@ public function testInvalidSecuredClosureWithoutSecuriyProvider()
-     */
     public function testInvalidJsonSecuredClosureWithoutSecuriyProvider()
     {
+        $this->expectException(SecurityException::class);
```
You'll need to do method exists checks so you can support the old versions of PHPUnit.

GrahamCampbell commented:
Thanks for the PR. I've left some feedback.

GrahamCampbell commented on tests/SignedClosureTest.php:
```diff
@@ -12,9 +12,14 @@
 class SignedClosureTest extends ClosureTest
 {
+    /**
+     * @expectedException \Opis\Closure\SecurityException
```
Please don't do this with the annotation. Call the legacy `setExpectedException` method.

GrahamCampbell commented on tests/SignedClosureTest.php:
```diff
@@ -115,9 +140,14 @@ public function testInvalidSecuredClosureWithoutSecuriyProvider()
     {
-        $this->expectException(SecurityException::class);
+        if (method_exists($this, 'expectException')) {
+            $this->expectException(SecurityException::class);
```
You can't use the ::class notation. We need to support PHP 5.4.

sorinsarca commented on src/ReflectionClosure.php:
```diff
@@ -84,27 +84,30 @@ public function getCode()
+            } elseif (version_compare(PHP_VERSION, '7.4', '<=')) {
+                $scalar_types = array('string', 'int', 'bool', 'float', 'void', 'object');
+            } else {
+                $scalar_types = array('string', 'int', 'bool', 'float', 'void', 'object', 'mixed');
```
not all types from here are scalars, the var name should probably be `$php_types`

GrahamCampbell replied:
Just call them `builtinTypes`. That's what PHP calls them.

GrahamCampbell commented:
@sorinsarca Is there any reason we don't include array and callable in the list of builtin types?

GrahamCampbell commented:
And also iterable?

sorinsarca commented:
@GrahamCampbell all should be there, I don't know why they are missing

GrahamCampbell commented:
// cc @driesvints for review

GrahamCampbell commented on src/ReflectionClosure.php:
```diff
@@ -84,27 +84,30 @@ public function getCode()
-                case 2:
-                default:
-                    $php7_types = array('string', 'int', 'bool', 'float', 'void', 'object');
+        if (version_compare(PHP_VERSION, '7.0.0', '>=')) {
```
I think you are viewing an old diff. This was already fixed.

GrahamCampbell commented on tests/SignedClosureTest.php:
```diff
@@ -18,7 +18,7 @@ class SignedClosureTest extends ClosureTest
     public function testSecureClosureIntegrityFail()
     {
         if (method_exists($this, 'expectException')) {
-            $this->expectException(SecurityException::class);
```
I think reverting can be done in the next PR, which actually removes support for PHP 5.

GrahamCampbell commented on tests/ReflectionClosure6Test.php:
```diff
@@ -0,0 +1,31 @@
+
+    public function testUnionTypes()
+    {
+        $f1 = fn() : string|int => 1;
```
Oh, I think we need to add `false` in to the builtin types array too...

GrahamCampbell replied:
Friendly ping @deleugpn. :)

sorinsarca replied:
We still need some tests with class names

deleugpn replied:
Let me know if that's what you expected.

sorinsarca replied:
Yes, but you can refer to a class in different ways:

```php
namespace Test;
use Some\ClassName as ClassAlias;

$x = function () : null | MyClass | ClassAlias | Relative\Ns\ClassName | \Absolute\Ns\ClassName {
  return null;
};
```

GrahamCampbell replied:
Ping @deleugpn. :)

deleugpn replied:
Sorry,  I have been on holiday for the last 2 weeks. I'll be back in 5 days. I don't have my laptop with me.

terabytesoftw commented:
hi, this will merge soon thanks, great Job.

deleugpn commented:
Sorry for not getting back to this sooner. I had some holidays to use up. I noticed it was not previously not running the `ReflectionClosure6Test` because of `phpunit.xml` configuration. I fixed that, but I'm having some issues with tests failing due to wrong Namespace. Locally all tests are passing. I'm trying to get to the bottom of why namespaces suddenly started failing.
I also added a couple more tests regarding Namespace requested by **sorinsarca** and I also rebased with current `master` branch.

deleugpn commented on phpunit.xml:
```diff
@@ -28,7 +28,7 @@
             <file phpVersion="7.4.0" phpVersionOperator=">=">./tests/ReflectionClosure5Test.php</file>
         </testsuite>
         <testsuite name="ReflectionClosure6">
-            <file phpVersion="8.0.0-x" phpVersionOperator=">=">./tests/ReflectionClosure6Test.php</file>
```
I specifically changed this because `8.0.0-dev` is not greater than `8.0.0`. The file is being completely ignored. On my local declaring it as `8.0.0-x` is correctly running the file

GrahamCampbell replied:
`8.0.0` should work. It works for mockery's test suite. https://github.com/mockery/mockery/blob/master/phpunit.xml.dist

deleugpn replied:
Apparently what makes it work for Mockery is `<directory suffix="Test.php">./tests</directory>` and not the phpVersion comparison line.

deleugpn replied:
See https://3v4l.org/VIgRZ

GrahamCampbell replied:
Hmmm. What if you wrote `8.0.0-dev`, there?

deleugpn replied:
Done

deleugpn commented:
Okay, so I have the `8.0.0-dev` identified by `PHP 8.0.0-dev (cli) (built: Jul 10 2020 16:12:58) ( NTS )`. GitHub Actions is running on `PHP 8.0.0-dev (cli) (built: Aug 12 2020 00:48:48) ( NTS )` which might explain why I'm getting all tests green and GitHub is failing a bunch.
I'm going to try and get the newest version on my machine, but given the fact that old tests are failing, I'd say there might be some regression on PHP?
ping @carusogabriel and @nikic

deleugpn commented:
I managed to get the same errors on my local and I started debugging this. The failing tests are due the fact that namespaces are not being correctly serialized/unserialized, possibly due to the namespace token change caused by the attribute `@@` RFC.  I'm still trying to debug and pinpoint exactly which part of the `$token` switch/case fails and what a fix would be.

GrahamCampbell commented:
The `@@` stuff will definitely be changing before the end of the month.

deleugpn commented:
The `@@` per se is not the important part, but the fact that it lead to https://wiki.php.net/rfc/namespaced_names_as_token which changed how `token_get_all()` works.

> Additionally tooling based on token_get_all() will need to be adjusted to handle the new T_NAME_* tokens. In practice, this will be the main negative impact of this proposal.

GrahamCampbell commented:
> https://wiki.php.net/rfc/namespaced_names_as_token 

Right, yeh. I think we could just add a BC layer which converts the new tokens to the old ones, tbh. That was my plan to get php-cs-fixer to work with PHP 8.0 class name stuff.

GrahamCampbell commented on src/ReflectionClosure.php:
```diff
@@ -910,6 +915,11 @@ protected function fetchItems()
                             $name .= $token[1];
                             $alias = $token[1];
                             break;
+                        case T_NAME_QUALIFIED:
```
There are more tokens than this to consider. ;)

deleugpn replied:
This one fixes one specific test that I wrote that was passing on 7.4. It's 23:00 here so I'm heading to bed for today and will continue tomorrow

GrahamCampbell replied:
This is still unresolved.

deleugpn replied:
Can you be more specific?

sorinsarca commented:
I have a backport of 4.x parser and I'll add it to a new branch tomorrow. I think we should add the php 8 support on top of this new parser since it is easier to extend/modify. 

However, the real problem is typed properties (starting with php 7.4) - see #45 - and so far this PR was only about the parser.

deleugpn commented:
Sorry for the delay, this was so much harder than I thought it would be. I finally got all existing tests passing. 

I added some new test files during the development process because it was much easier for me to understand the array of `tokens` with a smaller file. I was pretty much running it on PHP 7.4 and 8.0 and trying to identify different tokens. Let me know if I should revert the new phpunit test files.

I'm not sure what to do with these: https://github.com/opis/closure/pull/67/files#diff-4e781872947f7a06b71197212fc075bcR46
I only took inspiration from Nikita's parser to define the tokens when they don't exist.

Also let me know about the new parser, instead of causing conflict I can try to reimplement. If the new parser is easier to work with, I might be able to use some of the knowledge I gathered while working on this PR to get it working there.

In conclusion, my work here is done for now and this is ready for review

sorinsarca commented on src/ReflectionClosure.php:
```diff
@@ -36,6 +36,15 @@ class ReflectionClosure extends ReflectionFunction
      */
     public function __construct(Closure $closure, $code = null)
     {
+        // Not sure where to put this
```
I think these definitions can be moved before the class definition.

deleugpn replied:
Awesome idea. Done!

sorinsarca commented on src/ReflectionClosure.php:
```diff
@@ -143,7 +135,7 @@ public function getCode()
                         $code .= $token[1];
                         $state = $token[0] === T_FUNCTION ? 'function' : 'static';
-                    } elseif ($fn && $token[0] === T_FN) {
+                    } elseif (\PHP_VERSION_ID >= 70400 && $token[0] === T_FN) {
```
We should define `T_FN` before the class definition so we can remove the related version checks (`\PHP_VERSION_ID >= 70400`)
```php
defined('T_FN') || define('T_FN', -6);
```

sorinsarca commented on tests/ReflectionClosure6Test.php:
```diff
@@ -0,0 +1,48 @@
+
+        $this->assertEquals($e1, $this->c($f1));
+    }
+}
```
We also need test for PHP 8 attributes, nullsafe operator, named arguments, trailing comma - every syntax change of PHP 8 must be tested

sorinsarca commented on tests/NamespaceFullyQualifiedTest.php:
```diff
@@ -0,0 +1,22 @@
+
+final class NamespaceGroupTest extends \PHPUnit\Framework\TestCase
+{
+    public function test_namespace_fully_qualified()
```
More tests needed here

deleugpn replied:
This test was just a duplicate of tests already written in other files. The reason I copied it here was so that I would have less Tokens when debugging what was crashing and adding more cases to the parser.

deleugpn commented:
- [x] Nullsafe Operator
- [x] Trailing Comma
- [ ] Named Parameter
- [ ] Attributes
- [ ] Constructor Property Promotion

GrahamCampbell commented:
I think we can merge this without attributes working. A priority is getting PHP 7 compatible syntax to work on PHP 8, which this PR already does. A follow-up PR can add support for every PHP 8.0 feature (which is still a slightly moving target until RC1 comes out, anyway).

GrahamCampbell commented:
> A priority is getting PHP 7 compatible syntax to work on PHP 8

The lack of this is blocking Laravel from supporting PHP 8.0.

deleugpn commented:
Thanks! I'll try to look into Attributes now that we're in RC stage of PHP.

GrahamCampbell commented:
I've opened a separate issue to track this: https://github.com/opis/closure/issues/80. It is possible that we might only want to bother with attribute support in v4 of this package (after this PR has been re-implemented for v4). We can discuss on that new issue.

mnapoli commented:
Ohh yeah thank you for merging, I'm working on PHP 8 support in PHP-DI and that was a blocker.

Hey there @deleugpn, fun seeing you here! 😄

Blitheness commented:
Should the README be updated to mention PHP 8.0 support? At present, only PHP 5 and 7 are mentioned.
